lit_llama/utils.py

```diff
@@ -207,7 +207,7 @@ def __getattr__(self, name):
         if name in {"size"}:
             return getattr(self.metatensor, name)
         # materializing with contiguous is needed for quantization
-        if name in {"contiguous"}:
+        if name in {"contiguous", "as_subclass"}:
```

Comment:
@t-vi 
I added this to fix a loading error as seen by @rasbt https://github.com/Lightning-AI/lit-llama/pull/187#issuecomment-1522114427

Is this fix correct? I'm not really sure why it is needed. I am also not able to identify why this was even a problem. If I load the checkpoint and inspect it, it just only has regular torch Tensors in it, nothing unexpected.

Reply:
So this likely depends intimately on the PyTorch version and I am afraid that it is *not* the solution, as it would undo the lazy loading.
I think the right fix would involve diverting the `_rebuild_from_type_v2` but we'd need to check. @rasbt  What is the exact PyTorch version you are using?

Reply:
To make matters more tricky, do you know which PyTorch version created the archive?

Reply:
OK, so after some more research, this is because what is in the state dict are not Tensors themselves but a subclass. @rasbt do you know which one? At any rate, I'll see that I fix the loading for these.

Reply:
@t-vi This happens with the checkpoint created with pytorch 2.0 and loading it. The subclass call confused me because the tensors in the checkpoint are definitely not subclasses. But the dtype is bfloat16, and that seems to be the problem. If I convert the checkpoint to fp32 and lazy_load that, it works.

Reply:
Maybe the conversion changes the subclass? Out of curiosity, what is the class that gets passed when loading?
The reason I think that it is worth checking is that while lazy_load should just handle subclasses (I'm working on that ETA in 5 hours or so), if the deepspeed & co cause subclasses to be pickled, you need the frameworks to load the checkpoint, which seems suboptimal.

Reply:
Could this issue perhaps be related to how we save the finetuned model here? 

```python
from lit_llama.utils import save_model_checkpoint

save_model_checkpoint(fabric, model, os.path.join(out_dir, "lit-llama-full-finetuned.pth"))
```

I can try out using the regular

```python
fabric.save(os.path.join(out_dir, "lit-llama-full-finetuned.pth"), checkpoint)
```

Reply:
Could you maybe modify your installed PyTorch to print the class (`new_type`) passed to `_rebuild_from_type_v2`
https://github.com/pytorch/pytorch/blob/e5291e633fca9639ace4fb4585e1bb0b9dca9e95/torch/_tensor.py#L57 ? I would be very keen to see what is saved...

Reply:
Hope you don't mind intercepting and answering for @rasbt 😅 

Printing the new_type like you said in torch._tensor: _rebuild_from_type_v2 shows `<class 'torch.Tensor'>`. As I said, that is also what confused me. 

Note, it's the `NotYetLoadedTensor` that shows the attribute error of `as_subclass`:

```
Traceback (most recent call last):
  File "/home/adrian/repositories/lightning-llama/generate.py", line 160, in <module>
    CLI(main)
  File "/home/adrian/.conda/envs/lit-llama/lib/python3.10/site-packages/jsonargparse/cli.py", line 82, in CLI
    return _run_component(component, cfg_init)
  File "/home/adrian/.conda/envs/lit-llama/lib/python3.10/site-packages/jsonargparse/cli.py", line 138, in _run_component
    return component(**cfg)
  File "/home/adrian/repositories/lightning-llama/generate.py", line 118, in main
    checkpoint = lazy_load(checkpoint_path)
  File "/home/adrian/repositories/lightning-llama/lit_llama/utils.py", line 243, in lazy_load
    sd = mup.load()
  File "/home/adrian/.conda/envs/lit-llama/lib/python3.10/pickle.py", line 1213, in load
    dispatch[key[0]](self)
  File "/home/adrian/.conda/envs/lit-llama/lib/python3.10/pickle.py", line 1590, in load_reduce
    stack[-1] = func(*args)
  File "/home/adrian/.conda/envs/lit-llama/lib/python3.10/site-packages/torch/_tensor.py", line 61, in _rebuild_from_type_v2
    ret = ret.as_subclass(new_type)
  File "/home/adrian/repositories/lightning-llama/lit_llama/utils.py", line 213, in __getattr__
    raise AttributeError(f"{type(self)} does not have {name}")
AttributeError: <class 'lit_llama.utils.NotYetLoadedTensor'> does not have as_subclass
```

@t-vi I'm going to build a mini-repro script so it's easier to debug

Reply:
Thanks @awaelchli 

Btw it works perfectly fine with

```python
fabric.save(os.path.join(out_dir, "lit-llama-full-finetuned.pth"), checkpoint)
``` 

for the sake of simplicity/hackability of this script, this is maybe the better one to use here anyways? I'll update the PR

Reply:
> @t-vi I'm going to build a mini-repro script so it's easier to debug

Thank you so much! So the thing that really surprises me is that PyTorch records _rebuild_from_type_v2 when doing the pickling (rather than `_rebuild_tensor_v2`) when it  sees torch.Tensor, but it could be what gets passed to it...

Reply:
> Btw it works perfectly fine with
> fabric.save(os.path.join(out_dir, "lit-llama-full-finetuned.pth"), checkpoint)

Nevermind, this was accidentally the old `checkpoint`, when replacing this with `model.state_dict()` it's the same issue as above.

Reply:
@rasbt Yes, feel free to use that if it works. So I think this discussion here shouldn't block the PR, and we can fix this independently.

@t-vi As promised, here is the minimized pytorch repro:

```py
import torch
from lit_llama.utils import lazy_load
from torch.distributed.fsdp import FullStateDictConfig
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
from torch.distributed.fsdp import StateDictType

torch.distributed.init_process_group(
    init_method="tcp://localhost:12345", backend="nccl", world_size=1, rank=0
)
model = torch.nn.Linear(2, 2)
model = FSDP(model, device_id=0)
save_policy = FullStateDictConfig(offload_to_cpu=False, rank0_only=True)
with FSDP.state_dict_type(model, StateDictType.FULL_STATE_DICT, save_policy):
    state_dict = model.state_dict()

assert type(state_dict["weight"]) == torch.Tensor  # a regular tensor indeed!
torch.save(state_dict, "saved.pt")
lazy_load("saved.pt")
# torch.load("saved.pt")

```
The FSDP must be modifying the tensor somehow and causing _rebuild_from_type_v2 to be called.

Reply:
@rasbt How about we just use torch.load() for now in `generate_full.py` instead of lazy_load? I think this would be the best workaround. Fabric does not yet support proper full-state-dict saving (that's why I have the custom saving in the save_checkpoint utility).

Reply:
Tried that, but oh no, a different error now 😅

```python
File "/home/sebastian/Developer/lora/lit-llama/generate.py", line 118, in main
    checkpoint = lazy_load(checkpoint_path)
  File "/home/sebastian/Developer/lora/lit-llama/lit_llama/utils.py", line 243, in lazy_load
    sd = mup.load()
  File "/home/sebastian/miniforge3/envs/lit-llama/lib/python3.9/pickle.py", line 1212, in load
    dispatch[key[0]](self)
  File "/home/sebastian/miniforge3/envs/lit-llama/lib/python3.9/pickle.py", line 1589, in load_reduce
    stack[-1] = func(*args)
  File "/home/sebastian/miniforge3/envs/lit-llama/lib/python3.9/site-packages/torch/_tensor.py", line 60, in _rebuild_from_type_v2
    ret = ret.as_subclass(new_type)
  File "/home/sebastian/Developer/lora/lit-llama/lit_llama/utils.py", line 213, in __getattr__
    raise AttributeError(f"{type(self)} does not have {name}")
AttributeError: <class 'lit_llama.utils.NotYetLoadedTensor'> does not have as_subclass
(lit-llama) sebastian@hyperplane1:~/Developer/lora/lit-llama$ python generate.py --checkpoint_path out/full2/alpaca/lit-llama-full2-finetuned.pth 
Loading model ...
Traceback (most recent call last):
  File "/home/sebastian/Developer/lora/lit-llama/generate.py", line 160, in <module>
    CLI(main)
  File "/home/sebastian/miniforge3/envs/lit-llama/lib/python3.9/site-packages/jsonargparse/cli.py", line 82, in CLI
    return _run_component(component, cfg_init)
  File "/home/sebastian/miniforge3/envs/lit-llama/lib/python3.9/site-packages/jsonargparse/cli.py", line 147, in _run_component
    return component(**cfg)
  File "/home/sebastian/Developer/lora/lit-llama/generate.py", line 119, in main
    model.load_state_dict(checkpoint)
  File "/home/sebastian/miniforge3/envs/lit-llama/lib/python3.9/site-packages/torch/nn/modules/module.py", line 2041, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for LLaMA:
        Missing key(s) in state_dict: "lm_head.weight", "transformer.wte.weight", "transformer.ln_f.scale". 
        Unexpected key(s) in state_dict: "_flat_param". 
(lit-llama) sebastian@hyperplane1:~/Developer/lora/lit-llama$ python generate_full.py --checkpoint_path out/full2/alpaca/lit-llama-full2-finetuned.pth ^C
(lit-llama) sebastian@hyperplane1:~/Developer/lora/lit-llama$ python generate.py --checkpoint_path out/full2/alpaca/lit-llama-full2-finetuned.pth 
Loading model ...
Traceback (most recent call last):
  File "/home/sebastian/Developer/lora/lit-llama/generate.py", line 160, in <module>
    CLI(main)
  File "/home/sebastian/miniforge3/envs/lit-llama/lib/python3.9/site-packages/jsonargparse/cli.py", line 82, in CLI
    return _run_component(component, cfg_init)
  File "/home/sebastian/miniforge3/envs/lit-llama/lib/python3.9/site-packages/jsonargparse/cli.py", line 147, in _run_component
    return component(**cfg)
  File "/home/sebastian/Developer/lora/lit-llama/generate.py", line 119, in main
    model.load_state_dict(checkpoint)
  File "/home/sebastian/miniforge3/envs/lit-llama/lib/python3.9/site-packages/torch/nn/modules/module.py", line 2041, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for LLaMA:
        Missing key(s) in state_dict: "lm_head.weight", "transformer.wte.weight", "transformer.ln_f.scale". 
        Unexpected key(s) in state_dict: "_flat_param".
```